File: labellab-flask/api/helpers/log.py
from sqlalchemy import desc

from api.extensions import db
from api.models.Log import Log

# Fetch all logs related to a project
def fetch_all_project_logs(project_id):
  logs = Log.query.filter_by(project_id=project_id).order_by(desc('timestamp')).all()
  return list(map(lambda log: to_json(log), logs))

# Fetch all project logs of a particular category
def fetch_all_category_logs(project_id, category):
  logs = Log.query.filter_by(project_id=project_id, category=category).order_by(desc('timestamp')).all()
  return list(map(lambda log: to_json(log), logs))

# Fetch all project logs of a particular project member
def fetch_all_user_logs(project_id, user_id):
  logs = Log.query.filter_by(project_id=project_id, user_id=user_id).order_by(desc('timestamp')).all()
  return list(map(lambda log: to_json(log), logs))

# Fetch all logs of a particular entity in a project
def fetch_all_entity_logs(project_id, entity_type, entity_id):
  logs = Log.query.filter_by(
    project_id=project_id,
    entity_type=entity_type, 
    entity_id=entity_id,
  ).order_by(desc('timestamp')).all()
  return list(map(lambda log: to_json(log), logs))

# Fetch last 5 logs of a project
def fetch_recent_project_logs(project_id):
  logs = Log.query.filter_by(project_id=project_id).order_by(desc('timestamp')).limit(5)
  return list(map(lambda log: to_json(log), logs))

# Save a log to db
def save_log(log):
  db.session.add(log)
  db.session.commit()

# Delete all project logs
def delete_all_project_logs(project_id):
  Log.query.filter_by(project_id=project_id).delete()
  db.session.commit()

# Return log model as json
def to_json(log):
  return {
    'id': log.id,
    'message': log.message,
    'category': log.category,
    'entity_id': log.entity_id,
    'entity_type': log.entity_type,
    'user_id': log.user_id,
    'username': log.username,
    'project_id': log.project_id,
    'timestamp': log.timestamp,
  }

File: labellab_mobile/lib/model/log.dart
class Log {
  int? id;
  String? message;
  String? category;
  int? userId;
  String? username;
  int? projectId;
  String? timestamp;
  int? entityId;
  String? entityType;

  Log(
    int id,
    String message,
    String category,
    int userId,
    String username,
    int projectId,
    String timestamp,
    int entityId,
    String entityType,
  ) {
    this.id = id;
    this.message = message;
    this.category = category;
    this.userId = userId;
    this.username = username;
    this.projectId = projectId;
    this.timestamp = timestamp;
    this.entityId = entityId;
    this.entityType = entityType;
  }

  Log.fromJSON(Map<String, dynamic> data) {
    id = data['id'];
    message = data['message'];
    category = data['category'];
    userId = data['user_id'];
    username = data['username'];
    projectId = data['project_id'];
    timestamp = data['timestamp'];
    entityId = data['entity_id'] != null ? data['entity_id'] : null;
    entityType = data['entity_type'] != null ? data['entity_type'] : null;
  }

  Map<String, dynamic> toMap() {
    return {
      'id': this.id,
      'message': this.message,
      'category': this.category,
      'project_id': this.projectId,
      'user_id': this.userId,
      'username': this.username,
      'entity_type': this.entityType,
      'entity_id': this.entityId,
    };
  }
}
